File: CSC310-Python/Module10-Randomized-Algorithms/debug01/shuffle_tally.py
"""Uniform random permutation in place (RANDOMIZE-IN-PLACE, CLRS 5.3).

A correct shuffle must make every one of the n! permutations equally
likely, which implies that each element is equally likely to end up in
each position.  The main block shuffles the list ['A', 'B', 'C'] many
times with a fixed seed and tallies where 'A' lands and how often each
of the six permutations appears, so the distribution can be inspected.
"""

import random

rng = random.Random(2024)


def shuffle(items):
    n = len(items)
    for i in range(n):
        j = rng.randint(0, n - 1)
        items[i], items[j] = items[j], items[i]


def main():
    trials = 100_000
    base = ["A", "B", "C"]
    position_of_a = [0, 0, 0]
    permutations = {}
    for _ in range(trials):
        items = base[:]
        shuffle(items)
        position_of_a[items.index("A")] += 1
        key = "".join(items)
        permutations[key] = permutations.get(key, 0) + 1

    expected = trials / 3
    print(f"{trials} shuffles of {base}")
    print("where 'A' ends up:")
    for pos, count in enumerate(position_of_a):
        print(f"  position {pos}: {count:>6}   ({100 * count / trials:5.2f}%)")
    print(f"  expected about {expected:.0f} each; largest deviation "
          f"{max(abs(c - expected) for c in position_of_a):.0f}")
    print()
    print("permutation frequencies (expected about "
          f"{trials / 6:.0f} each):")
    for key in sorted(permutations):
        print(f"  {key}: {permutations[key]:>6}")
    spread = max(permutations.values()) - min(permutations.values())
    print(f"  spread between most and least common: {spread}")
    print(f"  uniform within 3%: {spread < 0.03 * trials / 6}")


if __name__ == "__main__":
    main()

File: CSC310-Cpp/_solutions/Module10-Randomized-Algorithms/debug01/shuffle_tally.cpp
// shuffle_tally.cpp
// Empirical check of an in-place random shuffle (RANDOMIZE-IN-PLACE, CLRS
// 5.3). The driver shuffles the sequence 0 1 2 many times with a fixed-seed
// generator and tallies how often each of the 6 permutations appears. A
// uniform shuffle should hit every permutation about equally often.
#include <iomanip>
#include <iostream>
#include <map>
#include <random>
#include <string>
#include <vector>

std::mt19937 rng(2024);   // fixed seed so every run gives the same tallies

// Shuffles a in place so that every permutation is equally likely.
void shuffle(std::vector<int>& a) {
    const int n = static_cast<int>(a.size());
    for (int i = n - 1; i > 0; --i) {
        std::uniform_int_distribution<int> pick(0, i);
        int j = pick(rng);
        std::swap(a[i], a[j]);
    }
}

int main() {
    const int trials = 108000;
    const int n = 3;
    std::map<std::string, int> tally;

    for (int t = 0; t < trials; ++t) {
        std::vector<int> a = {0, 1, 2};
        shuffle(a);
        std::string key;
        for (int v : a) {
            key += static_cast<char>('0' + v);
        }
        ++tally[key];
    }

    const double expected = static_cast<double>(trials) / 6.0;
    std::cout << "shuffles of " << n << " elements: " << trials
              << "   (expected per permutation: " << expected << ")\n";
    int worst = 0;
    for (const auto& [perm, count] : tally) {
        int diff = count - static_cast<int>(expected);
        worst = std::max(worst, std::abs(diff));
        std::cout << "  " << perm << ": " << std::setw(6) << count
                  << "  (" << std::showpos << diff << std::noshowpos << ")\n";
    }
    std::cout << "largest deviation from expected: " << worst << '\n';
    std::cout << "permutations seen: " << tally.size() << " of 6\n";
    return 0;
}
